File: save_quotes_in_markdown.py
# *_*coding:utf-8 *_*
# @Author: SZJ
from __future__ import absolute_import, division, print_function

import json
import random
import string
import shutil
from pathlib import Path


class QuotesSaver:
    def __init__(self, input_file_path):
        """
        初始化TextCleaner对象。

        :param input_file_path: 输入的JSON文件的路径。
        """
        self.input_file = Path(input_file_path)
        self.data = self._load_data()

    def _load_data(self):
        """
        从文件路径加载JSON数据。

        :return: 解析JSON后的数据。
        """
        with self.input_file.open(encoding='utf-8') as file:
            return json.load(file)

    def shuffle_quotes(self):
        """
        随机打乱引用数据的顺序。
        """
        random.shuffle(self.data)
        return self.data

    def generate_random_filename(self, length=8):
        """
        生成随机文件名。

        :param length: 随机字符串的长度，默认为8。
        :return: 随机生成的文件名。
        """
        letters = string.ascii_lowercase
        return '来源quotes_' + ''.join(
            random.choice(letters) for _ in range(length)) + '.md'

    def save_quotes(self, output_dir):
        """
        遍历self.data并将每个引用保存到单独的Markdown文件中。

        :param output_dir: 输出目录的路径。
        """
        output_path = Path(output_dir)
        output_path.mkdir(parents=True, exist_ok=True)

        for item in self.data:
            quote = item.get('quote', '')
            author = item.get('author', '')
            if quote:
                filename = 'quotes_' + '_'.join(quote[:12]
                .replace('。', '')
                .replace('；', '')
                .replace('：', '')
                .replace('[', '')
                .replace(']', '')
                .replace('.', '')
                .replace('《', '')
                .replace('》', '')
                .replace('"', '')
                .replace('"', '')
                .replace('，', '').split(
                    ' ')) + '.md'
                # filename = self.generate_random_filename()
                file_path = output_path / filename
                with file_path.open('w', encoding='utf-8') as file:
                    if '佚名' in author:
                        file.write(f"{quote}")
                    else:
                        file.write(f"{quote}\n\n--{author}")
                print(f"Quote saved to: {file_path}")

    def save_quotes_to_single_file(self, output_dir, shuffle=True):
        """
        将所有引用保存到单个quotes.md文件中。

        :param output_dir: 输出目录的路径。
        :param shuffle: 是否打乱引用顺序，默认为True。
        """
        output_path = Path(output_dir)
        output_path.mkdir(parents=True, exist_ok=True)

        file_path = output_path / 'quotes.md'

        # 如果需要打乱顺序
        if shuffle:
            self.shuffle_quotes()
            print("数据已随机打乱")

        with file_path.open('w', encoding='utf-8') as file:
            for item in self.data:
                quote = item.get('quote', '')
                author = item.get('author', '')
                if quote:
                    if '佚名' in author:
                        file.write(f"{quote}\n\n\n")
                    else:
                        file.write(f"{quote}\n--{author}\n\n\n")

        print(f"All quotes saved to: {file_path}")

        # 复制cover.jpg到quotes.md所在目录
        cover_src = Path('src/assets/cover.jpg')
        cover_dst = output_path / 'cover.jpg'

        try:
            shutil.copy2(cover_src, cover_dst)
            print(f"Cover image copied to: {cover_dst}")
        except FileNotFoundError:
            print(f"警告: 封面图片文件 {cover_src} 未找到")
        except Exception as e:
            print(f"复制封面图片时出错: {e}")


if __name__ == '__main__':
    source_dir = Path.home() / 'Downloads' / f'quotes'
    source_dir.mkdir(exist_ok=True, parents=True)

    quotes_saver = QuotesSaver('src/assets/quotes.json')
    quotes_saver.save_quotes_to_single_file(source_dir, shuffle=True)
    print('well done!!!')

File: save_quotes_in_markdown.ts
import * as fs from 'fs';
import * as path from 'path';
import * as shutil from 'fs-extra';
// Import epub-gen using require for compatibility
const Epub = require('epub-gen');

interface QuoteItem {
    quote: string;
    author: string;
}

class QuotesSaver {
    private input_file: string;
    private data: QuoteItem[];

    constructor(input_file_path: string) {
        /**
         * 初始化QuotesSaver对象。
         *
         * @param input_file_path 输入的JSON文件的路径。
         */
        this.input_file = input_file_path;
        this.data = this._load_data();
    }

    private _load_data(): QuoteItem[] {
        /**
         * 从文件路径加载JSON数据。
         *
         * @return 解析JSON后的数据。
         */
        const fileContent = fs.readFileSync(this.input_file, 'utf-8');
        return JSON.parse(fileContent);
    }

    public shuffle_quotes(): QuoteItem[] {
        /**
         * 随机打乱引用数据的顺序。
         */
        this.data.sort(() => Math.random() - 0.5);
        return this.data;
    }


    public async save_quotes_to_epub(output_dir: string, title: string = "温故知心", author: string = "尼古拉西格玛宋", shuffle: boolean = true): Promise<void> {
        /**
         * 将所有引用保存为EPUB电子书格式。
         *
         * @param output_dir 输出目录的路径。
         * @param title 电子书标题
         * @param author 电子书作者
         * @param shuffle 是否打乱引用顺序，默认为True。
         */
        const output_path = path.resolve(output_dir);
        fs.mkdirSync(output_path, {recursive: true});

        const epub_path = path.join(output_path, `${title}.epub`);

        // 如果需要打乱顺序
        if (shuffle) {
            this.shuffle_quotes();
            console.log("数据已随机打乱");
        }

        // 准备EPUB内容
        const content = [];
        const chapters = [];

        // 将引用分组，每组n条作为一章
        const quotesPerChapter = 100;
        let chapterCount = 0;
        let currentChapter = [];

        for (const item of this.data) {
            const quote = item.quote || '';
            const author = item.author || '';

            if (quote) {
                currentChapter.push({ quote, author });

                if (currentChapter.length >= quotesPerChapter) {
                    chapterCount++;
                    chapters.push({
                        title: `第${chapterCount}章`,
                        data: [...currentChapter]
                    });
                    currentChapter = [];
                }
            }
        }

        // 处理最后一章（如果有剩余引用）
        if (currentChapter.length > 0) {
            chapterCount++;
            chapters.push({
                title: `第${chapterCount}章`,
                data: [...currentChapter]
            });
        }

        // 为每一章生成HTML内容
        for (const chapter of chapters) {
            let chapterContent = "";
            // let chapterContent = `<h1>${chapter.title}</h1>\n`;

            for (const item of chapter.data) {
                chapterContent += `<div class="quote-item">
                    <p class="quote">${item.quote}</p>
                    <p class="author">---- ${item.author}</p>
                    <br><br><br>
                </div>\n`;
            }

            content.push({
                title: chapter.title,
                data: chapterContent
            });
        }

        // 准备封面图片路径
        const cover_src = path.join('src', 'assets', 'cover.jpg');
        let coverPath = null;

        try {
            if (fs.existsSync(cover_src)) {
                coverPath = cover_src;
            }
        } catch (error) {
            console.log(`警告: 获取封面图片时出错: ${error.message}`);
        }

        // 创建EPUB选项
        const options = {
            title: title,
            author: author,
            publisher: 'QuotesSaver',
            cover: coverPath,
            content: content,
            output: epub_path,  // 添加输出路径
            css: `
                .quote-item {
                    margin-bottom: 2em;
                    page-break-inside: avoid;
                }
                .quote {
                    font-size: 1.0em;
                }
                .author {
                    text-align: right;
                    font-style: italic;
                }
            `
        };

        try {
            // 生成EPUB文件
            await new Epub(options).promise;
            console.log(`EPUB电子书已保存至: ${epub_path}`);
        } catch (err) {
            console.error(`生成EPUB时出错: ${err.message}`);
        }
    }
}

// 主函数
async function main(): Promise<void> {
    const homedir = require('os').homedir();
    const source_dir = path.join(homedir, 'Downloads', 'quotes');
    fs.mkdirSync(source_dir, {recursive: true});

    const quotes_saver = new QuotesSaver('src/assets/quotes.json');

    // 保存为EPUB电子书
    try {
        await quotes_saver.save_quotes_to_epub(source_dir, "温故知心");
        console.log('well done!!!');
    } catch (error) {
        console.error('生成EPUB失败:', error);
    }
}

// 执行主函数
main().catch(error => {
    console.error('程序执行出错:', error);
});


